standard/lpdb.lua

```diff
@@ -0,0 +1,60 @@
+---
+-- @Liquipedia
+-- wiki=commons
+-- page=Module:Lpdb
+--
+-- Please see https://github.com/Liquipedia/Lua-Modules to contribute
+--
+
+local Lpdb = {}
+
+local _DEFAULT_QUERY_LIMIT = 20
+local _DEFAULT_INITIAL_OFFSET = 0
+
+--[==[
+Wrapper for mass LPDB queries.
+Used when > 5000 results are needed (LPDB max limit is 5000) or when
+additional filtering is to be done after the query and memory is an issue.
+
+example:
+	local cond = '[[match2id::!]]'
+	local query = 'match2id'
+
+	local function queryFunct(foundElements, offset, maxQueryLimit)
+		local data = mw.ext.LiquipediaDB.lpdb('match2', {
+			limit = maxQueryLimit,
+			offset = offset,
+			conditions = cond,
+			query = query,
+		})
+
+		for _, item in ipairs(data) do
+			if string.match(item.match2id, '_0001') then
+				table.insert(foundElements, item.match2id)
+			end
+		end
+
+		return foundElements, #data
+	end
+
+	local queryData = Lpdb.massQueryWrapper(queryFunct, 0, 5000)
+]==]
+function Lpdb.massQueryWrapper(funct, initialOffset, maxQueryLimit, maxRounds)
+	local offset = initialOffset or _DEFAULT_INITIAL_OFFSET
+	local count = maxQueryLimit or _DEFAULT_QUERY_LIMIT
+	maxRounds = maxRounds or math.huge
+	local rounds = 0
+
+	local foundElements = {}
+
+	while count == maxQueryLimit and rounds < maxRounds do
+		foundElements, count = funct(foundElements, offset, maxQueryLimit)
+
+		offset = offset + maxQueryLimit
+		rounds = rounds + 1
+	end
+
+	return foundElements
+end
```

Comment:
```lua
function Lpdb.executeQuery(lpdbTable, queryParameters, callbackFunction, limit)
	queryParameters.offset = queryParameters.offset or 0
	queryParameters.limit = queryParameters.limit or 5000

	local lpdbData = {}
	while queryParameters.offset < limit then
		queryParameters.limit = Math.min(queryParameters.limit, limit - queryParameters.offset)

		lpdbData = mw.ext.LiquipediaDB.lpdb(lpdbTable, queryParameters)
		Table.iter.forEachIndexed(lpdbData, callbackFunction)

		queryParameters.offset = queryParameters.offset + #lpdbData
		if #lpdbData < queryParameters.limit then
			break
		end
	end
end
```

Reply:
And on the calling side of things
```lua
local placementData = {}
local processPlacement = function(index, placement)
	-- Ommited: Process the row somehow
	-- Saving it for later
	table.insert(placementData, placement)
end
local query = {
	conditions = '[[pageid::!-1]] AND [[participant::!Tbd]] AND [[participant::!TBD]] AND [[placement::!]]',
	query = 'date, participant, players, prizemoney, placement, liquipediatier, liquipediatiertype, mode',
}
Lpdb.executeQuery('placements', query, processPlacement)
```

Reply:
- that would error, because the call doesn't have the necessary limit variable set
- it would not allow breaking the loop from a condition in the `callbackFunction`
- - there are cases where you just do not want to condition on a limit but on the number of found entries (i already have a sc2 module that needs that)

Reply:
> * that would error, because the call doesn't have the necessary limit variable set

Yes I wrote a comment below the implementation about it

>     * it would not allow breaking the loop from a condition in the `callbackFunction`

Easy to add support

>     * * there are cases where you just do not want to condition on a limit but on the number of found entries (i already have a sc2 module that needs that)

Do you mean the overall limit, or the per lpdb call limit? Either way, both are supported.

Reply:
maybe
```lua
function Lpdb.executeQuery(lpdbTable, queryParameters, callbackFunction, limit, breakCallbackFunction)
	queryParameters.offset = queryParameters.offset or 0
	queryParameters.limit = queryParameters.limit or 5000
	breakCallbackFunction = breakCallbackFunction or Lpdb._defaultBbreakCallbackFunction

	local lpdbData = {}
	while queryParameters.offset < limit do
		queryParameters.limit = Math.min(queryParameters.limit, limit - queryParameters.offset)

		lpdbData = mw.ext.LiquipediaDB.lpdb(lpdbTable, queryParameters)
		Table.iter.forEachIndexed(lpdbData, callbackFunction)

		queryParameters.offset = queryParameters.offset + #lpdbData
		if #lpdbData < queryParameters.limit or breakCallbackFunction() then
			break
		end
	end
end

function Lpdb._defaultBbreakCallbackFunction()
	return false
end
```

the call then could look like
```lua
local foundMatchIds = {}
local processMatch = function(index, match)
	if #foundMatchIds < args.matchLimit then
		foundMatchIds = HeadToHead._extractMatchesByAdditionalConditions(foundMatchIds, match, args)
	end
end
local breakFunction = function()
	return #foundMatchIds < args.matchLimit
end
local queryParameters = {
	conditions = conditions,
	order = 'date ' .. args.order
	query = 'pagename, winner, walkover, finished, date, dateexact, links, '
		.. 'bestof, vod, tournament, tickername, shortname, icon, icondark, '
		.. 'extradata, match2opponents, match2games, mode, match2id, match2bracketid',
}
Lpdb.executeQuery('match2', queryParameters, processMatch, math.huge, breakFunction)
```

Reply:
```lua
function Lpdb.executeQuery(lpdbTable, queryParameters, callbackFunction, limit)
	queryParameters.offset = queryParameters.offset or 0
	queryParameters.limit = queryParameters.limit or 5000
	limit = limit or math.huge

	local lpdbData = {}
	while queryParameters.offset < limit then
		queryParameters.limit = Math.min(queryParameters.limit, limit - queryParameters.offset)

		lpdbData = mw.ext.LiquipediaDB.lpdb(lpdbTable, queryParameters)
		for _, value in ipairs(lpdbData) do
			local success = callbackFunction(value)
			if success == false then
				return
			end
		end

		queryParameters.offset = queryParameters.offset + #lpdbData
		if #lpdbData < queryParameters.limit then
			break
		end
	end
end
```

Reply:
changed it to that and tested it already